docs/protocol/extension_trace.md

```diff
@@ -0,0 +1,193 @@
+# EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+# IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+# PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL THE COPYRIGHT OWNER OR
+# CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+# EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+# PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+# PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY
+# OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+# (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+# OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+-->
+
+# Trace Extension
+
+This document describes Triton's trace extension. The trace extension enables
+the client to configure the trace settings during a Triton run. Because this
+extension is supported, Triton reports “trace” in the extensions field of
+its Server Metadata.
+
+## HTTP/REST
+
+In all JSON schemas shown in this document $number, $string, $boolean,
+$object and $array refer to the fundamental JSON types. #optional
+indicates an optional JSON field.
+
+Triton exposes the trace endpoint at the following URL. The client may use
+HTTP GET request to retrieve the current trace setting. A HTTP POST request
+will modify the trace setting, and the endpoint will return the updated trace
+setting on success or an error in the case of failure. Optional model name
+can be provided to get or to set the trace settings for specific model.
+Note that when a model is specified without providing any trace settings
+in the request body, it will be a request to remove the specification. 
+
+```
+GET v2[/models/${MODEL_NAME}]/trace
+
+POST v2[/models/${MODEL_NAME}]/trace
+```
+
+### Trace Response JSON Object
+
+A successful trace request is indicated by a 200 HTTP status
+code. The response object, identified as $trace_response, is
+returned in the HTTP body for every successful trace request.
+
+```
+$trace_response =
+{
+  $trace_setting, ...
+}
+
+$trace_setting = $string : $string | [ $string, ...]
+```
+
+Each $trace_setting JSON describes a “name”/”value” pair, where the “name” is
+the name of the trace setting and the “value” is a $string representation of the
+setting value, or an array of $string for some settings. Currently the following
+trace settings are defined:
+
+- "trace_file" : the file where the trace output will be saved. If
+"log frequency" is set, this will be the prefix of the files to save the
+trace output, resulting files in name "${trace_file}.0", "${trace_file}.1"...,
+see trace setting "log frequency" below for detail.
+- "trace_level" : the trace level. "OFF" to disable tracing,
+TIMESTAMPS" to trace timestamps, "TENSORS" to trace tensors.
+This value is an array of string whhere user may specify multiple levels to
+trace multiple informations.
+- "trace_rate" : the trace sampling rate. The value represents how many requests
```

Comment:
Need to add "trace_count" here and also in the cmdline. trace_count defaults to 0 which means to continue collecting trace indefinitely. But it can be set to non-zero to cause trace_level to go to OFF after collecting trace_count traces.

Reply:
If `trace_level` is updated to OFF when `trace_count` exceeds, I assume the following will be true?
1. Turning OFF global `trace_level` by `trace_count` does not cause any model specific setting to turn off its `trace_level` even if it wasn't specified for the model. In other words, the `trace_level` in model setting will be marked as "specified" once the global `trace_level` is affected by `trace_count`
2. The user can turn `trace_level` back on with the update API. For global setting, due to the behavior in 1., this change will not be reflected to model setting unless the user "clear" the model `trace_level` first.

Reply:
Would it be easier to explain and implement if we don't actually change trace_level when trace_count is done? In the implementation we would trace only if trace_level was not OFF and trace_count > 0. We can use trace_count = -1 to mean unlimited trace_count.

Also be easier for user to just reset trace_count if they want to restart or change it (no need to also reset trace_level). Also, we can be consistent withe trace_count, when it is set in global and has not been specifically set in model then we can just set it in the model as well.

Reply:
So we use `trace_count` to directly indicate the remaining trace to be sampled for the particular setting, and a -1 value is used for no trace limit? Yes, that will be more clear to know why exactly the trace stops.

Reply:
Yes, and we don't adjust trace_level ourselves.